core/trino-spi/src/main/java/io/trino/spi/block/IntArrayBlockEncoding.java

```diff
@@ -81,14 +87,70 @@ public IntArrayBlock readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput
         int positionCount = sliceInput.readInt();
 
         byte[] valueIsNullPacked = retrieveNullBits(sliceInput, positionCount);
-        int[] values = new int[positionCount];
-
         if (valueIsNullPacked == null) {
+            int[] values = new int[positionCount];
             sliceInput.readInts(values);
             return new IntArrayBlock(0, positionCount, null, values);
         }
-        boolean[] valueIsNull = decodeNullBits(valueIsNullPacked, positionCount);
+        else {
+            boolean[] valueIsNull = decodeNullBits(valueIsNullPacked, positionCount);
+            if (vectorizeNullExpand) {
+                return expandIntsWithNullsVectorized(sliceInput, positionCount, valueIsNull);
+            }
+            else {
+                return expandIntsWithNullsScalar(sliceInput, positionCount, valueIsNullPacked, valueIsNull);
+            }
+        }
+    }
 
+    private static IntArrayBlock expandIntsWithNullsVectorized(SliceInput sliceInput, int positionCount, boolean[] valueIsNull)
+    {
+        if (valueIsNull.length != positionCount) {
+            throw new IllegalArgumentException("valueIsNull length must match positionCount");
+        }
+        int nonNullPositionCount = sliceInput.readInt();
+        int nonNullIndex = positionCount - nonNullPositionCount;
+        int[] values = new int[positionCount];
+        sliceInput.readInts(values, nonNullIndex, nonNullPositionCount);
+
+        int position = 0;
+        if ((nonNullPositionCount * (INT_SPECIES.length() * 4L)) <= positionCount) {
```

Comment:
Do we get a significant gain from doing separate branchy vs branchless versions ?
So far we have stuck to using one or the other in all the hot code paths as it keeps a lid on the code complexity even if its not the most optimized implementation in every case. 
I think it would be nicer to retain just the branchless version and accept that it is slightly slower on the many NULLs path (which should be the less common case typically).

Reply:
Unfortunately yes, the regression compared to scalar is in the ~20% range compared to scalar code even on CPU's with the most complete intrinsic support for vectorization. The issue is most apparent with `long` (but also visible with `int`) and stems from the number of extra value loads that result in no non-null stores being incurred over and over again (hence why the penalty is highest for `long` and why this extra logic is omitted for `byte` and `short` where no regression is visible).